src/network/socket.cpp

```diff
@@ -0,0 +1,478 @@
+Minetest
+Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "socket.h"
+
+#include <stdio.h>
+#include <iostream>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <sstream>
+#include <iomanip>
+#include "util/string.h"
+#include "util/numeric.h"
+#include "constants.h"
+#include "debug.h"
+#include "settings.h"
+#include "log.h"
+
+#ifndef _WIN32
+	#include <sys/types.h>
+	#include <fcntl.h>
+	#include <netdb.h>
+	#include <unistd.h>
+	#include <arpa/inet.h>
+#endif
+
+// Windows defines setsockopt with a char* fourth parameter instead of the
+// POSIX void*, but Cygwin fixes this.
+#if defined(_WIN32) && !defined(__CYGWIN__)
+	#define setsockopt(a, b, c, d, e) \
+		setsockopt(a, b, c, reinterpret_cast<const char*>(d), e)
+#endif
+
+// Set to true to enable verbose debug output
+bool socket_enable_debug_output = false;
+
+
+// Windows needs sockets to be initialized before use
+#ifdef _WIN32
+static bool g_sockets_initialized = false;
+
+static inline void cleanup_sockets() { WSACleanup(); }
+
+static inline void init_sockets()
+{
+	WSADATA data;
+	if (WSAStartup(MAKEWORD(2, 2), &data) != NO_ERROR)
+		throw SocketException("Failed to initialize Windows sockets!");
+	atexit(cleanup_sockets);
+	g_sockets_initialized = true;
+}
+#endif
+
+
+
+// Address
+
+// Equality (address family, address and port must be equal)
+bool Address::operator == (const Address &other)
+{
+	if (other.family != family || other.port != port)
+		return false;
+
+	switch (family) {
+	case AF_INET: return address.v4.s_addr == other.address.v4.s_addr;
+	case AF_INET6: return memcmp(address.v6.s6_addr,
+			other.address.v6.s6_addr, 16) == 0;
+	default:
+		return false;
+	}
+}
+
+void Address::resolve(const std::string &name)
+{
+	if (name.empty()) return setZero();
+
+	struct addrinfo *resolved, hints = {0};
+
+	// Setup hints
+	if (g_settings->getBool("enable_ipv6")) {
+		// AF_UNSPEC allows both IPv6 and IPv4 addresses to be returned
+		hints.ai_family = AF_UNSPEC;
+	} else {
+		hints.ai_family = AF_INET;
+	}
+
+	// Do getaddrinfo()
+	int err = getaddrinfo(name.c_str(), NULL, &hints, &resolved);
+	if (err)
+		throw ResolveError(gai_strerror(err));
+
+	// Copy data
+	switch (resolved->ai_family) {
+	case AF_INET: {
```

Comment:
Shouldn't bodies inside of cases be like...
switch (foo) {
case FOO_BAR: {
    <sdfgh>;
    break;
}
case FOO_BAZ: {
 ...
?